### ISSUE TITLE
How/where to install ESD kernel module 

### ISSUE THREAD

**Issue by lyletbjohnson** [ApolloAuto/apollo-kernel#13]
After building the ESD kernel module with the custom Makefile, it's unclear where to copy the `esdcan-pcie402.ko` file.

I'm following instructions from [this guide](https://github.com/ApolloAuto/apollo-kernel/blob/master/linux/ESDCAN-README.md), category "Build ESD Kernel Driver with Pre-Built Apollo Kernel", Step 5.

How to you install the kernel module to the `apollo_dev` and/or `apollo_release` Docker containers?

**Comment by lyletbjohnson** [comment]
I noticed that the kernel module may be missing when the ESD example application was executed from the Docker container, and an error is emitted  indicating the `*.so` library cannot be located. After searching the filesystem, I could not locate any kernel modules or development files.

**Comment by startcode** [comment]
Hi, you can try to use the `insmod` command to load the *.ko file to kernel.
```
sudo insmod /path/to/your/esdcan-pcie402.ko
```

Here is a reference document http://www.tldp.org/LDP/lkmpg/2.6/html/x44.html

If you can find esd can driver from `lsmod` like the following command, you should be fine.
```
lsmod | grep -i esd
```

In addition, we will update the document to make it more clear.

**Comment by lyletbjohnson** [comment]
The `esdcan-pcie402.ko` file does not exist on the target; I need help understanding how to properly install it to the target. 

The kernel module was properly built against the `apollo-kernel` on the host machine, and currently exists in the host machines file system.

**Comment by startcode** [comment]
If I understand correctly, you can build esdcan-pcie402.ko by yourself following this guide.
 https://github.com/ApolloAuto/apollo-kernel/blob/master/linux/ESDCAN-README.md

After build, you can load `esdcan-pcie402.ko` with the `insmod` command on the host machine, not in docker.

**Comment by lyletbjohnson** [comment]
I realize I made some incorrect assumptions yesterday. 

Please correct me if I'm wrong:

1. The `esdcan-pcie402.ko` kernel module does not need to exist on the target; it needs to be loaded by the kernel on the host machine
2. When the `esdcan-pcie402.ko` module is loaded by the host system kernel, the module can be detected with the `lsmod` tool
3. The target will see `/dev/can*` devices enumerated when the host system properly detects the PCIe CAN interface through the ESD driver

---

When I attempt to use `insmod` to load the load the module, I receive the following error:

```
insmod: ERROR: could not insert module ./esdcan-pcie402.ko: No such device
```

I've ensured that the PCIe card is properly installed / seated well. Are there other common causes to this error message?

**Comment by startcode** [comment]
Can you firs try use command `lscpi` to see if the PCIe card has been correctly installed and can be detected in the system?

`lspci | grep -i esd`

The `insmod` command won't work unless `lspci` shows the PCIe card has been detected.

If possible, you many want to try other PCIe slot in the computer

**Comment by lyletbjohnson** [comment]
Thanks for your suggestion. I was able to detect the PCIe card with `lspci`. 

Next I re-downloaded the prebuilt Apollo kernel, cleaned the ESD source code directory, and rebuilt the kernel module. 

After rebooting the machine I was able to detect the loaded kernel module with `lsmod`. 

My issue must have been an incompatible kernel. Thank you for your help, and for improving the documentation for future users.
